### ISSUE TITLE
not able to connect to workflow in workflow inputs

### ISSUE THREAD

**Issue by dmaticzka** [galaxyproject/galaxy#1739]
Hi, I was checking out the new workflow in workflow tool from 16.01 and found that I can only connect "basic" workflow inputs to inputs of this tool. It does not connect if the datatype is changed, e.g. to tabular by the cut tool.

Steps to reproduce:
- check out fresh 16.01
- create a workflow consisting of one input and one concatenate "single_input_cat"

![wf_single_input_cat](https://cloud.githubusercontent.com/assets/113329/13077869/c8501280-d4bb-11e5-8182-68452651dd90.png)
- create new workflow with single input and try to connect to "single_input_cat"
  - first row: _can't_ connect via cut
  - second row: **can** connect directly from input
  - third row: **can** connect via cat
  - fourth row: **can** connect cut and cat

![test_wf_in_wf](https://cloud.githubusercontent.com/assets/113329/13077959/61b9874e-d4bc-11e5-854e-843aa55fe15b.png)

**Comment by jmchilton** [comment]
:scream: Yeah - I suppose. I'll try to find time for a fix to this - the functionality is clearly heavily limited without this.

**Comment by hexylena** [comment]
Similar issue reported in a bullet point of #1387

**Comment by hexylena** [comment]
Still experiencing this @jmchilton, 17.05. (test_a just contains an `input dataset → require_txt` step, require_txt inputs + outputs a txt format file.)

Input datasets work fine,
![2017-06-15-194627_1366x768_scrot](https://user-images.githubusercontent.com/458683/27198932-b5d76b64-5203-11e7-96a1-dabf051af5ff.png)

Other tools do not.
![2017-06-15-194632_1366x768_scrot](https://user-images.githubusercontent.com/458683/27198938-b8038332-5203-11e7-9aee-39933ac085db.png)

**Comment by hexylena** [comment]
[test.xml.txt](https://github.com/galaxyproject/galaxy/files/1078817/test.xml.txt)
[Galaxy-Workflow-test_b.ga.txt](https://github.com/galaxyproject/galaxy/files/1078819/Galaxy-Workflow-test_b.ga.txt)
[Galaxy-Workflow-test_a.ga.txt](https://github.com/galaxyproject/galaxy/files/1078818/Galaxy-Workflow-test_a.ga.txt)

Please find attached some test cases.

I'm using the following patch to work around this (the functionality is TOO valuable for us)

```diff
diff --git a/lib/galaxy/workflow/modules.py b/lib/galaxy/workflow/modules.py
index fd65716..c8c5c1e 100644
--- a/lib/galaxy/workflow/modules.py
+++ b/lib/galaxy/workflow/modules.py
@@ -282,7 +282,7 @@ class SubWorkflowModule( WorkflowModule ):
                     name=name,
                     label=name,
                     multiple=False,
-                    extensions="input",
+                    extensions=["input", "data"],
                     input_type=step_to_input_type[step_type],
                 )
                 inputs.append(input)
```

**Comment by nsoranzo** [comment]
We see this too with @anilthanki on Galaxy 17.09.

**Comment by jmchilton** [comment]
Oh crap - I must have been swamped because I didn't see this until now - I'm so sorry. Yes I'd say that patch is perfect it just doesn't go far enough - it should be "data" and not ["input", "data"] - that "input" stuff there now is quite broken and would only allow input modules to connect in this way.

**Comment by nsoranzo** [comment]
Fixed by #5140.
